src/core/ext/client_channel/client_channel_factory.c

```diff
@@ -85,3 +90,120 @@ grpc_arg grpc_client_channel_factory_create_channel_arg(
   arg.value.pointer.vtable = &factory_arg_vtable;
   return arg;
 }
+
+struct grpc_channel_targets_info {
```

Comment:
Yes, mdstr_hash_table was designed specifically for mdstr keys (hence the name), so it would require either creating mdstrs for the keys or changing the mdstr_hash_table implementation to use different key types.

The gpr_avl code might be a better choice, though.  It looks to me like it would work as-is.

In any case, I'm fine with doing this after you're done merging changes from head.